### PR TITLE
[FA] Alibi Forward Pass

### PR DESCRIPTION
**Pull request by micmelesse**
* this is the forward pass for adding alibi to our FA Kernel

### CODE REVIEW

File: python/perf-kernels/flash-attention.py [hunk 1]
```diff
@@ -953,8 +1007,8 @@ def backward(ctx, do, _):
 
 @pytest.mark.parametrize('Z, H, N_CTX_Q, N_CTX_K, D_HEAD',
                          [(4, 48, 1024, 1024, 64),
-                          (4, 48, 8192, 8192, 64),
-                          (2, 16, 16384, 16384, 128),
+                          (1, 24, 8192, 8192, 64),
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:1010]: Were the old sizes too large?

File: python/perf-kernels/flash-attention.py [hunk 2]
```diff
@@ -963,14 +1017,15 @@ def backward(ctx, do, _):
...
-                          (1, 8, 16330, 15989, 128),
+                          (1, 4, 16330, 15989, 128),
                           (4, 4, 1024, 1024, 33),
                           (4, 4, 65, 1019, 65),
                           (4, 4, 128, 128, 65),
                           (4, 4, 113, 123, 1),
                           ])
 @pytest.mark.parametrize('causal', [False, True])
-def test_op_fwd(Z, H, N_CTX_Q, N_CTX_K, D_HEAD, causal, dtype=torch.float16):
+@pytest.mark.parametrize('use_alibi', [True])
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:1027]: Can you also add False so we can test both?

File: python/perf-kernels/flash-attention.py [hunk 3]
```diff
@@ -979,6 +1034,13 @@ def test_op_fwd(Z, H, N_CTX_Q, N_CTX_K, D_HEAD, causal, dtype=torch.float16):
     if causal:
         input_metadata.need_causal()
 
+    if use_alibi:
+        # for n heads the set of slopes is the geometric sequence that starts 2^(-8/n)
+        alibi_slopes = torch.tensor([2**(-8/H*i) for i in range(1, H+1)], dtype=torch.float32, device="cuda").repeat(Z, 1)
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:1039]: I haven't read the paper thoroughly. OOC, Is this just a test value you used, or does the paper actually recommend this sequence?

File: python/perf-kernels/flash-attention.py [hunk 4]
```diff
@@ -996,6 +1058,13 @@ def test_op_fwd(Z, H, N_CTX_Q, N_CTX_K, D_HEAD, causal, dtype=torch.float16):
         mask = torch.tril(torch.ones(N_CTX_Q, N_CTX_K, device="cuda"), 
                           diagonal=N_CTX_K-N_CTX_Q)
         scores[:, :, mask==0] = float("-inf")
+    if use_alibi:
+        q_idx = torch.arange(N_CTX_Q, dtype=torch.int32, device="cuda").unsqueeze(-1)
+        k_idx = torch.arange(N_CTX_K, dtype=torch.int32, device="cuda").unsqueeze(0)
+        relative_pos = torch.tril(k_idx - q_idx)
+        alibi = alibi_slopes.unsqueeze(-1).unsqueeze(-1) * relative_pos # (Z, H, N_CTX_Q, N_CTX_K)
+        scores+= alibi
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:1066]: Hmm so the documentation on Tri Dao's implementation says 

```
alibi_slopes: (nheads,) or (batch_size, nheads), fp32. A bias of
            (-alibi_slope * |i + seqlen_k - seqlen_q - j|)
            is added to the attention score of query i and key j.
```

So the value it must be multiplied by would be `torch.abs(q_idx - k_idx + K_CTX - Q_CTX)`. But here instead of abs it is `max(0, q_idx - k_idx)` from what I can tell. I'm probably missing something?

File: python/perf-kernels/flash-attention.py (hunk 3 above)
Comment (micmelesse) [reply python/perf-kernels/flash-attention.py:1039]: yes, this is from the paper

File: python/perf-kernels/flash-attention.py (hunk 1 above)
Comment (micmelesse) [reply python/perf-kernels/flash-attention.py:1010]: yes

File: python/perf-kernels/flash-attention.py (hunk 4 above)
Comment (micmelesse) [reply python/perf-kernels/flash-attention.py:1067]: thanks, I fixed it.

File: python/perf-kernels/flash-attention.py [hunk 5]
```diff
@@ -387,6 +420,19 @@ def attn_fwd(
         )
     else:
         bias_ptr = None
+
+    if USE_ALIBI != 0:
+        alibi_slopes_ptr = tl.make_block_ptr(
+            base=alibi_slopes,  # The base pointer to the parent tensor
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:426]: A bit nitpicky, but as we will eventually upstream this and it will be representing AMD flash attention, a minor code style suggestion would be to move these comments up to the first block pointer declaration. A reader can assume the others follow the same concept.

File: python/perf-kernels/flash-attention.py [hunk 6]
```diff
@@ -387,6 +420,19 @@ def attn_fwd(
         )
     else:
         bias_ptr = None
+
+    if USE_ALIBI != 0:
+        alibi_slopes_ptr = tl.make_block_ptr(
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:425]: I think we don't need a block pointer for this. IIUC, for a given head / batch, the entire QK result uses the same scalar alibi value which would mean each instance of this kernel is loading a scalar. We can use an old school `tl.load(offset)` for that.

File: python/perf-kernels/flash-attention.py [hunk 7]
```diff
@@ -202,6 +212,24 @@ def _attn_fwd_inner(
             # our optimization to use 2^x instead of e^x results in an additional
             # scale factor of log2(e) which we must also multiply the bias with.
             qk += (bias * 1.44269504089)
+           
+        if alibi_slopes_ptr is not None:
+            alibi_slopes_block = load_fn(alibi_slopes_ptr, False, MASK_STEPS and (n_extra_tokens != 0), "zero")
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:217]: I don't think it is possible to go out of bounds for this. So you can use False, False for the bounds check args, or provide no mask in the tl.load (depending on if you take my suggestion to not use a block pointer for this or not).

File: python/perf-kernels/flash-attention.py (hunk 5 above)
Comment (micmelesse) [reply python/perf-kernels/flash-attention.py]: The notes where mostly for me. I removed them.

File: python/perf-kernels/flash-attention.py (hunk 6 above)
Comment (micmelesse) [reply python/perf-kernels/flash-attention.py]: Done

File: python/perf-kernels/flash-attention.py (hunk 7 above)
Comment (micmelesse) [reply python/perf-kernels/flash-attention.py]: I am using just a tl.load at the start of the kernel.

**Review by vgokhale — APPROVED** [review]
